### PR TITLE
[#IR-294] Update the Assign Points algorithm

### PR DESCRIPTION
**Pull request by himanshu-garg-165**
#### Update the Assign Points algorithm to apply:
1. Just on the tasks which are marked done at the time of re-sync.
2. When a task is marked done by the user (asynchronously).

### CODE REVIEW

File: apps/retrospective/services/sprint_sync.go [hunk 1]
```diff
@@ -295,6 +295,7 @@ func (service SprintService) AssignPoints(sprintID string) (err error) {
 		Not("sprint_tasks.id in (?)", sprintTaskToSkipPointsAllocation).
 		Scopes(retroModels.NotDeletedSprint).
 		Where("tasks.retrospective_id = ?", sprint.RetrospectiveID).
+		Not("tasks.done_at IS NULL").
```
Comment (anmolghosh) [inline apps/retrospective/services/sprint_sync.go]: replace this with 
`where done_at is not null`

File: apps/retrospective/services/sprint_sync.go [hunk 2]
```diff
@@ -303,19 +304,22 @@ func (service SprintService) AssignPoints(sprintID string) (err error) {
...
+	updateSQL := `UPDATE sprint_member_tasks
 	    SET points_assigned = COALESCE(sprint_member_tasks.time_spent_minutes
 				/ NULLIF(s1.sprint_task_total_time_spent, 0) * s1.remaining_points, 0),
             points_earned = COALESCE(sprint_member_tasks.time_spent_minutes
 				/ NULLIF(s1.sprint_task_total_time_spent, 0) * s1.remaining_points, 0),
             updated_at = NOW()
         FROM (?) AS s1 
-        WHERE s1.sprint_id = ? and sprint_member_tasks.id = s1.id
-    `, dbs, sprintID).Error
+        WHERE s1.sprint_id = ? and sprint_member_tasks.id = s1.id`
```
Comment (anmolghosh) [inline apps/retrospective/services/sprint_sync.go]: `and` should be in caps

File: apps/retrospective/services/sprint_sync.go [hunk 3]
```diff
@@ -303,19 +304,22 @@ func (service SprintService) AssignPoints(sprintID string) (err error) {
...
+
+	if sprintTaskID != nil {
+		updateSQL = fmt.Sprintf("%v AND s1.sprint_task_id = %v", updateSQL, *sprintTaskID)
```
Comment (anmolghosh) [inline apps/retrospective/services/sprint_sync.go]: why are we using `%v`? Is `%s` usable?
Also don't append `sprintTaskID` here, that may lead to query injection. Append that in line 322, I think gorm should ignore extra params, if not you need to have two separate calls

**Review by anmolghosh — CHANGES_REQUESTED** [review]

File: apps/retrospective/services/sprint_sync.go (hunk 1 above)
Comment (himanshu-garg-165) [reply apps/retrospective/services/sprint_sync.go]: It is already with a `Not`!!

Comment (anmolghosh) [reply apps/retrospective/services/sprint_sync.go]: Agreed, but I'm not confident/comfortable with gorm's `Not`. 
`Not('column_name','value')` is okay but `Not("tasks.done_at IS NULL")` feels weird.
Can you share the actual query which is created when we write this statement

File: apps/retrospective/services/sprint_sync.go [hunk 4]
```diff
@@ -303,19 +304,25 @@ func (service SprintService) AssignPoints(sprintID string) (err error) {
...
 				/ NULLIF(s1.sprint_task_total_time_spent, 0) * s1.remaining_points, 0),
             points_earned = COALESCE(sprint_member_tasks.time_spent_minutes
 				/ NULLIF(s1.sprint_task_total_time_spent, 0) * s1.remaining_points, 0),
             updated_at = NOW()
         FROM (?) AS s1 
-        WHERE s1.sprint_id = ? and sprint_member_tasks.id = s1.id
-    `, dbs, sprintID).Error
+        WHERE s1.sprint_id = ? AND sprint_member_tasks.id = s1.id`
+
+    sqlValues := []interface{}{dbs, sprintID}
```
Comment (anmolghosh) [inline apps/retrospective/services/sprint_sync.go]: let's rename `dbs` to something more readable

**Review by anmolghosh — CHANGES_REQUESTED** [review]

File: Gopkg.lock [hunk 5]
```diff
@@ -4,20 +4,20 @@
 [[projects]]
   name = "cloud.google.com/go"
   packages = ["compute/metadata"]
-  revision = "777200caa7fb8936aed0f12b1fd79af64cc83ec9"
-  version = "v0.24.0"
+  revision = "64a2037ec6be8a4b0c1d1f706ed35b428b989239"
```
Comment (anmolghosh) [inline Gopkg.lock]: why was this file updated?

File: Gopkg.lock [hunk 6]
```diff
@@ -248,7 +248,7 @@
   branch = "master"
   name = "github.com/qor/admin"
   packages = ["."]
-  revision = "138548ee992b3b70c9e75e81f9aaa1744d49adfe"
+  revision = "c8f683ff9e646a1653dfd4bbbe2d9eb56bbc0b6a"
```
Comment (anmolghosh) [inline Gopkg.lock]: is this validated?

File: apps/retrospective/services/sprint_sync.go (hunk 1 above)
Comment (himanshu-garg-165) [reply apps/retrospective/services/sprint_sync.go]: Done

**Review by anmolghosh — DISMISSED** [review]

**Review by anmolghosh — APPROVED** [review]

**Pull request merged**